### PR TITLE
[crypto] Support for native threshold signature verification

### PR DESCRIPTION
**Pull request by kchalkias**
## Motivation

Support for _k out-of n_ threshold signature verification as a native function. This is essentially a concatenated signatures approach that provides multi-sig functionality to Libra, similarly to Bitcoin. Current implementation supports up to 32 public keys (and signatures).

## Algorithm
**Inputs:** 
- _n_ concatenated public keys, where `1 <= n <= 32`
- _k_ concatenated signatures, where `1 <= k <= n`
- the signed message
- a 32-bit `bitmap` (the number of its high set bits should be equal to `k`)

Signatures should be submitted in the correct order based on the public key ordering. The `bitmap` guides us on which signature corresponds to which key. For instance, in the following example the 1st signature is checked against the 1st public key (with index 0), the 2nd signature is checked against the 3rd  public key (with index 2), the 3rd signature is checked against the  4th public key (with index 3) and so on.

image
Also, the scheme does not allow failed signatures even if the threshold is met, thus all submitted  signatures should be valid. This is to prevent malicious users from adding arbitrary content to the blockchain masqueraded as invalid signatures.

### Have you read the [Contributing Guidelines on pull requests](https://github.com/libra/libra/blob/master/CONTRIBUTING.md#pull-requests)?

YES
## Test Plan
```
threshold_signature_ed25519_1_out_of_1.mvir
threshold_signature_ed25519_1_out_of_2.mvir
threshold_signature_ed25519_2_out_of_2.mvir
threshold_signature_ed25519_7_out_of_10.mvir
```

### CODE REVIEW

File: language/stdlib/natives/src/signature.rs [hunk 1]
```diff
@@ -31,3 +41,138 @@ pub fn native_ed25519_signature_verification<T: StackAccessor>(
...
+
+    let mut bitmap_last_bit_set: usize = 0; // This is fine as we expect at least one set bit.
+    let mut bitmap_count_ones: usize = 0;
+    for (i, bit) in bitmap.iter().enumerate() {
+        if bit {
+            bitmap_count_ones += 1;
+            bitmap_last_bit_set = i;
+        }
+    }
+    if bitmap_count_ones == 0 {
```
Comment (kchalkias) [inline language/stdlib/natives/src/signature.rs]: @sblackshear I specifically check the all zeros case.

**Comment by kchalkias** [comment]
btw @sblackshear what is a good pattern to create functional tests that we expect to fail? Ie if I wanted to test for all zeros etc? I could in theory return `U64(0)`, but I wanted to avoid that, I prefer an error type return.

**Comment by sblackshear** [comment]
You can add `// check:` directives that pattern-match against the expected error. Here is one [example](https://github.com/libra/libra/blob/master/language/functional_tests/tests/testsuite/payments/cant_recreate_market_cap.mvir).

There is a broader problem with returning errors from natives... Move doesn't have an equivalent of  `Result`, so it's probably best for us to avoid failing in natives as much as possible. Planning to write a post about this tomorrow.

**Comment by kchalkias** [comment]
Thanks @sblackshear! Expect an upcoming PR with extra functional tests for various types of failures.

**Pull request merged**